### PR TITLE
Cargo console ID check

### LINKED ISSUE #4242: Cargo computers should need an ID check, like the sec computers
## Description

Not only will that prevent abuse, it will also make the game more interesting for antags and make the "Who ordered 6 crates of laser rifles" questions answerable

Just a minor issure

### PR DESCRIPTION
**Pull request by Nukeuler123**
### Purpose
Fixes #4242 
Allows the cargo console to check for specific IDs. (QM, Caption, HOP, CargoTech, NULL, AI, Cyborg)

### Notes:
In Pogstation there is a custom gameobject called Cargo Console instead of supply console, it has been left alone for mappers to fix and in case changing a component on it breaks something. Will make issue.

### CODE REVIEW

**Comment by corp-0** [comment]
image
it always says invalidID for me

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+public class CargoConsole : NetworkBehaviour,ICheckedInteractable<HandApply>
+{
+	private bool correctID;
+	public bool CorrectID => correctID;
+
+	private GUI_Cargo associatedTab;
+
+	[SerializeField]
+	private List<JobType> allowedTypes = new List<JobType>();
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
	private List<JobType> allowedTypes = null;
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+	public bool WillInteract(HandApply interaction, NetworkSide side)
+	{
+		if (!DefaultWillInteract.Default(interaction, side))
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
		return DefaultWillInteract.Default(interaction, side) &&
		       Validations.HasComponent<IDCard>(interaction.HandObject);
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+			return false;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs:22]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		//interaction only works if using an ID card on console
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+		if (!Validations.HasComponent<IDCard>(interaction.HandObject))
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+			return false;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs:23]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+		return true;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs (hunk 3 above)
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
```

**Review by corp-0 — COMMENTED** [review]
just suggestions, its on you if you want to apply these

**Comment by corp-0** [comment]
```
NullReferenceException: Object reference not set to an instance of an object
CargoConsole.CheckID (JobType usedID) (at Assets/Scripts/Consoles/CargoConsole.cs:43)
```
 I get this when clicking the console with ID in hand on Fallstation.

```c
	private void CheckID(JobType usedID)
	{
		foreach (var aJob in allowedTypes.Where(aJob => usedID == aJob))
		{
			correctID = true;
			associatedTab.UpdateId();
			break;
		}
	}
```
associatedTab is null

**Comment by corp-0** [comment]
Tested in Outpost. It doesn't throw any NRE but it also doesn't do anything when I click console with ID in hand.

**Comment by Nukeuler123** [comment]
Alright, im pretty sure those errors are just due to some of the gameobjects on the maps, as for the suggestions ill commit em

**Comment by Nukeuler123** [comment]
Ill also begin more testing on different maps to see what I need to change

**Comment by corp-0** [comment]
the intended steps to take are clicking the cnsole with the ID and then click again without it to see the UI?

**Comment by Nukeuler123** [comment]
Should work either way, you can click on the console with the ID while in the UI or click on it with the ID without the UI, it stores the last ID it scanned and will reset when you buy something

**Comment by corp-0** [comment]
I'm thinking we might need a swipe interaction. How it is right now it is kind of hard to understand from the user perspective.

Something like cutting wires in the hacking GUI, but this time with the ID card. You click ON the GUI with the item in hand and that has an instant feedback in the same GUI.

**Comment by Bod9001** [comment]
is it ok to merge?

**Comment by corp-0** [comment]
At least in my test it didn’t work like intended

**Comment by Nukeuler123** [comment]
Seems to work fine, should I add some chat text for when someone uses their ID on the cargo console? Just for some visual feedback

**Comment by corp-0** [comment]
Yes, action messages are good. I will test again now and report back with a gif if something weird happens.

**Comment by corp-0** [comment]
GIF 08-08-2020 19-02-52
When I click the console with the ID nothing happens. The button still says "InvalidID" but I am able to make the request.

**Comment by Nukeuler123** [comment]
Hmm, after putting some debug log statements in my code... The code that should change the text does run... but apparently the text never updates for some reason...

**Comment by Nukeuler123** [comment]
should be fixed, i also added action messages

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+		// Not "optimized" for readability
+		if (correctID)
+		{
+			Chat.AddActionMsgToChat(playeref, "You swipe your ID through the supply console's ID slot, t" +
+			                                  "he console accepts your ID",
+				"swiped their ID through the supply console's ID slot");
+		}
+		else
+		{
+			Chat.AddActionMsgToChat(playeref, "You swipe your ID through the supply console's ID slot, t" +
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]: Break the string before the “t”, so the next string has the entire word

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+			                                  "he console denies your ID",
+				"swiped their ID through the supply console's ID slot");
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]: Missing the character name here. You can format the string like this
```
$”{playeref.ExpensiveName()} swiped their ID blah blah blah”
```
Also you should be able to get the possessive of the character’s gender but I’m not quite sure where is the function at the moment.

**Review by corp-0 — COMMENTED** [review]
Just gave it a quick look since I’m already in bed. Will test tomorrow!

**Comment by corp-0** [comment]
Tested, it works wonders. If you don't want to do the last suggestions I can commit them directly and then merge.

**Comment by Nukeuler123** [comment]
Sure go ahead, sorry for late response.

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs (hunk 10 above)
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
			Chat.AddActionMsgToChat(playeref, "You swipe your ID through the supply console's ID slot, " +
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+		if (correctID)
+		{
+			Chat.AddActionMsgToChat(playeref, "You swipe your ID through the supply console's ID slot, t" +
+			                                  "he console accepts your ID",
+				"swiped their ID through the supply console's ID slot");
+		}
+		else
+		{
+			Chat.AddActionMsgToChat(playeref, "You swipe your ID through the supply console's ID slot, t" +
+			                                  "he console denies your ID",
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
			                                  "the console denies your ID",
```

File: UnityProject/Assets/Scripts/Consoles/CargoConsole.cs (hunk 11 above)
Comment (corp-0) [inline UnityProject/Assets/Scripts/Consoles/CargoConsole.cs]:
```suggestion
				$"{playeref.ExpensiveName()} swiped their ID through the supply console's ID slot");
```

**Review by corp-0 — APPROVED** [review]

**Pull request merged**